Added #2353: Add ability to tie locations to companies - 2023 edition

Toreg87 opened the pull request:
# Description

Locations are the last big part of the application that can't be tied to companies.
This can be a problem with FullMultipleCompanySupport, because you can't restrict the visibility of locations to the company of the users.

The first commit wires everything up so that locations have a company_id and are properly scoped with FullMultipleCompanySupport.

The second commit introduce a setting to preserve backward compatibility with the existing behaviour.

This is an updated version of PR #10705. It is rebased to the current develop branch and a few bugs are fixed. I'm quite happy with the backward compatibility setting, so that erveryone can select the behaviour they want.

Fixes #2353

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

I checked that the correct locations are shown as SuperUser and scoped user with FullMultipleCompanySupport. 
The commits are tested with and without the backward compatibility setting.

# Checklist:

- [x] I have read the Contributing documentation available here: https://snipe-it.readme.io/docs/contributing-overview
- [x] I have formatted this PR according to the project guidelines: https://snipe-it.readme.io/docs/contributing-overview#pull-request-guidelines
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

Toreg87 committed:
Add ability to tie locations to companies

Locations are the last big part of the application that can't be tied to companies.
This can be a problem with FullMultipleCompanySupport, because you can't restrict the visibility of locations to the company of the users.

In order to change this, add a company_id to the locations table and wire everything up in the views and controllers.
Aditionally add a new formatter to filter the locations to a specific company, like it is done for assets.

Locations are properly scoped to the users company if FullMultipleCompanySupport is enabled.
If a parent location of a location has a different company than the user, the location does not show up.

Toreg87 committed:
Add a backward compatibility setting for locations with companies

Now that locations have a company_id they get restricted to the users company with FullMultipleCompanySupport.
This breaks backward compatibility, because before everyone can handle locations without restrictions.
Add a setting right below FullMultipleCompanySupport so that everyone can switch to the desired behaviour.
The default is off and the existing behaviour is preserved.

Toreg87 commented:
I found another bug in not using the CompanyableTrait in the location model. This complicates the backward compatibility a bit, but does now work correctly. I pushed the changes, this is now fixed.

hogemo commented:
Will this be available in the next version of Snipe IT? I really look forward for this change to be applied :)

snipe commented:
@paegelow not in the next version, but it's not off the table for merging. Again, this requires extensive testing as it can introduce security issues, and we're currently working on closing out the v6.x series to make way for v7.

Toreg87 commented:
@snipe I can help to test these changes and make adjustments if this is desired. Looking at Issue #2353 I think there is demand for this feature.

hogemo commented:
> @paegelow not in the next version, but it's not off the table for merging. Again, this requires extensive testing as it can introduce security issues, and we're currently working on closing out the v6.x series to make way for v7.

Hi again!

Could you provide a tentative timeline or any updates on when this might be reviewed for merging, especially considering the transition to v7? I understand if it's challenging to provide an exact date, but any insights would be greatly appreciated.

Thank you!

hogemo commented:
@Toreg87 Hi! Perhaps a stupid question, but how would i implement this in the most efficient way? Manually editing my files, and copy pasteing the changes?

Or, some git command? (i've done that before with not so much luck, so i ended up copy/pasteing, which also worked, but now with 17 files involved it seems a bit tedious..)

marcusmoore commented on database/migrations/2023_02_27_092130_add_scope_locations_setting.php:
```diff
@@ -0,0 +1,32 @@
+     */
+    public function down()
+    {
+        Schema::table('locations', function (Blueprint $table) {
```
This should reference the `settings` table.

Toreg87 replied:
Fixed

marcusmoore commented on database/migrations/2023_02_27_092130_add_scope_locations_setting.php:
```diff
@@ -0,0 +1,32 @@
+    public function up()
+    {
+        Schema::table('settings', function (Blueprint $table) {
+            $table->boolean('scope_locations_fmcs')->default('0');
```
This is a nitpick but can you group this column with `full_multiple_companies_support` please?
```suggestion
            $table->boolean('scope_locations_fmcs')->default('0')->after('full_multiple_companies_support');
```

Toreg87 replied:
Ok, changed.

marcusmoore commented on resources/views/settings/general.blade.php:
```diff
@@ -58,7 +58,22 @@
+                                {{ trans('admin/settings/general.scope_locations_fmcs_support_help_text') }}
+                            </p>
+                        </div>
+                    </div>
```
Can you update the mark up so the label is next to the input instead of below it ?
<img width="561" alt="Setting page" src="https://github.com/snipe/snipe-it/assets/1141514/7db9581c-d294-4487-bb41-b824c27edddb">

I think you need to wrap it in a `label.form-control`

Toreg87 replied:
Fixed

marcusmoore commented on app/Http/Controllers/SettingsController.php:
```diff
@@ -341,6 +341,13 @@ public function postSettings(Request $request)
         $setting->full_multiple_companies_support = $request->input('full_multiple_companies_support', '0');
+        $setting->scope_locations_fmcs = $request->input('scope_locations_fmcs', '0');
+
+        // Backward compatibility for locations makes no sense without FullMultipleCompanySupport
+        if (!$setting->full_multiple_companies_support) {
+            $setting->scope_locations_fmcs = '0';
+        }
```
This makes sense but might confuse some users. I wonder if there is a way to concisely express this on the settings page? 🤔  // @snipe

Toreg87 replied:
I think this can be handled in the view with some javascript. If this is a problem I can think about it some more, but my experience with this is quite limited.

marcusmoore commented on app/Models/Location.php:
```diff
@@ -208,6 +223,17 @@ public function parent()
+    */
+    public function company()
+    {
+        return $this->belongsTo('\App\Models\Company', 'company_id');
```
Just to match the style of surrounding methods:
```suggestion
        return $this->belongsTo(\App\Models\Company::class, 'company_id');
```

Toreg87 replied:
Fixed

marcusmoore commented on app/Http/Controllers/Api/LocationsController.php:
```diff
@@ -47,11 +63,17 @@ public function index(Request $request)
             ->withCount('rtd_assets as rtd_assets_count')
             ->withCount('users as users_count');
 
+        // Only scope locations if the setting is enabled
+        if (Setting::getSettings()->scope_locations_fmcs) {
+            $locations = Company::scopeCompanyables($locations);
+        }
```
Should this be wrapped in the following like we have in other places?

```
// Only scope the location if the setting is enabled
if (Setting::getSettings()->scope_locations_fmcs) {
    $location->company_id = Company::getIdForCurrentUser($request->input('company_id'));
} else {
    $location->company_id = $request->input('company_id');
}
```

Toreg87 replied:
This is in line with other scoped controllers and not about the company_id or am I misunderstanding something?

marcusmoore commented on app/Models/Location.php:
```diff
@@ -17,6 +18,16 @@
 
 class Location extends SnipeModel
 {
+    function __construct() {
```
I get the intent but I'd like to see if there is another way to approach this. Can this be handled in the `booted` method?

marcusmoore reviewed (changes requested):
Thanks for your work on this @Toreg87 😄 

I have some minor requests. 

Also, I want to figure out if we can move away from using the Location's constructor for adding the scope as well. I'll put some thought into that.

Would you be ok with me pushing tests to your branch?

Toreg87 commented:
@marcusmoore I'm currently on vacation until september. After that I will look at your requested changes. The constructor looks awfull, but this was the only way I could find to implement this. Suggestions for better alternatives are highly welcome. You can take the branch and push tests to it, no problem.

marcusmoore commented:
Note for myself/whoever is interested: The addition of the `__construct` method breaks the location factory so we can't seed or run tests in the current state.

I'm still working through this.

Toreg87 committed:
Another slightly less ugly way for backward compatibility

Instead of using a constructor, add a special check in the boot-method for locations.
This seems to fit better in the system and does hopefully not break the existing tests.

Signed-off-by: Tobias Regnery <tobias.regnery@gmail.com>

Toreg87 commented:
@marcusmoore I rebased my branch and fixed or commented your requested changes. I found another slightly less ugly way for the backward compatibility without the constructor. What do you think about it? I left the initial approach in the branch for discussion and can delete it later, if desired.

marcusmoore commented:
@Toreg87 thanks for updating and addressing the comments. I haven't had a chance to dive back into this but will as soon as I can.

hogemo commented:
Is there any update on this case? I have alot of users that are eager to try this :)

marcusmoore commented:
@paegelow this hasn't fallen off of my radar but unfortunately I haven't had a chance to get back to it yet.

dcdamien commented:
@marcusmoore is there something we can do to make this feature merged? This is really important for us.

marcusmoore commented:
@dcdamien thanks for your interest in this feature. It is still on my radar but I haven't had a chance to focus on it again but hope to in the near future. I have follow up on @Toreg87's changes before I can say if there other things to address before it is merged.

I've moved this higher up on my priority list and hope to dive back in soon.

Archi-web3 commented:
@marcusmoore Thank you very much Marcus for working on this feature, which is very important for many people, and I'll be keeping a close eye on progress. It's the only thing that has so far restricted us from using Snipe IT (I work for an international organization present in over 25 countries that would be interested in using Snipe IT). Is there anything we can do to help you move this forward? Provide a developer? or something. I know this might be hard to estimate, but do you have any idea about the timing of seeing this feature in a new version?

AdrianOsica commented:
Hi, @marcusmoore  What is your workload like? Do you already know the expected time for completion of this merge?

dcdamien commented:
@marcusmoore it seems your radar is rather long range one. In a little over 2 months, it will be appropirate to create new merge request, 2025 edition. Any chances to review it before end of the year?

snipe commented:
@dcdamien being a jerk to the devs isn’t going to get this merged faster. This is a dangerous PR as it inherently changes previous behavior that people have come to expect and rely on and there are merge conflicts and tests are failing. More flies with honey than vinagre and all that.

dcdamien commented:
> @dcdamien being a jerk to the devs isn’t going to get this merged faster. 

@snipe Well, note that my comment, unlike yours, does not offend anyone. I think you took it too seriously. Anyway, I think we won't focus on that, let's get to the specifics.

> This is a dangerous PR as it inherently changes previous behavior that people have come to expect and rely

Doesn't the MR address this issue by providing backward compatibility? Can you please explain what exacly is dangerous?
IMO the current behaviour is more dangerous as locations are leaking between organizations when the `Full Multiple Companies Support` is enabled.

> on and there are merge conflicts 

This is nominal that stale MR begins to conflict with the upstream branch, right?

> and tests are failing.

This MR is that old that I even can't check what's wrong with CI.

snipe commented:
> Well, note that my comment, unlike yours, does not offend anyone. I think you took it too seriously. Anyway, I think we won't focus on that, let's get to the specifics.

If I didn't think your comment was offensive, I wouldn't have said anything. 

> I think you took it too seriously.

I do take these things seriously, especially when people are being rude to my developers. The snide remark isn't useful.

> Doesn't the MR address this issue by providing backward compatibility? 

Yes, which adds additional test points and cyclomatic complexity, and potential complications with data down the line. 

> This is nominal that stale MR begins to conflict with the upstream branch, right?

If the OP can resolve those conflicts, we can take another look. A lot of things have changed since the original PR was opened, and guessing intent can be difficult with so many conflicts.

Toreg87 commented:
I try to find time to rebase this in the next couple of weeks.

dreamer2038 commented:
> I try to find time to rebase this in the next couple of weeks.

thank you for your hard work

protocol6v commented:
This feature is something that is stopping my company from deploying this to our end users. Its a critical privacy concern for us as an upstream company that "end users", who we need at various other 3rd party company locations to manage their assets, could view our entire locations list across all companies. 

Would really _really_ like to see this implemented soon.

Toreg87 commented:
@snipe I rebased my branch to the current version, can someone please have a look and review this?

One pain point is the parent-child-relationship of locations. With scoped locations and a parent with different company the relationship obviously isn't visible for scoped users. I don't know if this is a problem. For me it isn't, if I can scope my locations I don't need a parent location. But maybe others have different needs. Maybe we could introduce a bulk option and error out if there are different companies involved. 

But before I invest more time in this I think it would be good to get an agreement that this feature is useful und acceptable.

marcusmoore commented:
@Toreg87 

> But before I invest more time in this I think it would be good to get an agreement that this feature is useful und acceptable.

I completely agree and really appreciate your effort. Please give us a time to discuss this internally to make sure it is the right approach for the majority of users. I'd love to see this merged in but we went to be extra cautious it doesn't cause issues down the line.

dcdamien commented:
@snipe @marcusmoore please comment on @Toreg87 work.

swift2512 commented:
Aren't you overcomplicating everything? :) Departments are attached to companies, so why locations can't use same logic? **This would be first working step in right direction.** Can't imagine how binding location to company could impact users who don't use multiple companies. (And if they use multi company support, they probably hate current location system.) Multiple Companies using same location sounds odd, but even in this case nobody prevents you from creating another location with exact same name (ID will be different anyway). People from Company A can't see people, assets, departments in Company B so what changes if we add locations to that pool? I believe, nothing.

Currently, like others stated, locations are leaking to every single company (when Multi-Company is enabled) and anyone from any company can edit/delete those locations if have right permissions. (Actually, such behaviour is a GDPR breach.)

**Second step** - tool that displays a location tree for parent location and it's children. This tree could be put in the separate tab on 'location view' and it could be displayed like this:
```
PARENT (1)
↳Child (1.1)
  ↳Child (1.1.1)
  ↳Child (1.1.2)
    ↳Child (1.1.2.1)
  ↳Child (1.2)
  ↳Child (1.3
```
**Third step** - option to tie user to location where user or admin can't see users, items and locations higher then their assigned parent location (It's called 'scoped location'?). This alone removes the need to enable multi-company in the settings, because you can have different companies as different location trees. You would enable Multi-Company support if you really need people from different companies to use same Snipe-IT install.

Toreg87 commented:
Well maybe, but different comapnies have different needs. I wanted to make sure that this is a general solution for everyone. We can discuss the implementation when there is consensus with the devs that this feature can be merged.

snipe committed:
Merge branch 'develop' into feature/locations_with_companies

Toreg87 committed:
Check for inconsistencies before activating scoped locations

Before activating scoped location all locations and their related objects will be checked.
If there are locations with different companies than the related objects error out.

Because this operation is quite slow, bail out on the first inconsistent entry.
There is a new artisan command introduced that checks every location.

Depending on the size of the database, this will take very long.

Signed-off-by: Tobias Regnery <tobias.regnery@gmail.com>

Toreg87 committed:
Furhter validation for scoped locations

There is a new validator introduced that checks on object update (assets, users, etc.) if the company matches the locations company.
In case of the creation of a new location it must be checked that the parent matches the own company.
On updating a location a check for every related object must be made to see if the company matches the location.

Signed-off-by: Tobias Regnery <tobias.regnery@gmail.com>

snipe committed:
Merge branch 'develop' into feature/locations_with_companies
